src/sympc/tensor/replicatedshare_tensor.py

```diff
@@ -209,6 +249,138 @@ def ne(self, y):
+            return shares
+
+        return sum(shares)
+
+    @staticmethod
+    def __reconstruct_malicious(
+        share_ptrs: List["ReplicatedSharedTensor"],
+        get_shares: bool = False,
+    ) -> Union[torch.Tensor, List[torch.Tensor]]:
+        """Reconstruct value from shares.
+
+        Args:
+            share_ptrs (List[ReplicatedSharedTensor]): List of RSTensor pointers.
+            get_shares (bool): Retrieve only shares.
+
+        Returns:
+            reconstructed_value (torch.Tensor): Reconstructed value.
+
+        Raises:
+            ValueError: When parties share values are not equal.
+        """
+        nparties = len(share_ptrs)
+
+        # Get shares from all parties
+        all_shares = []
+        for party_rank in range(nparties):
+            all_shares.append(share_ptrs[party_rank].get_shares().get())
+
+        # reconstruct shares from all parties and verify
+        value = None
+        for party_rank in range(nparties):
+            share_sum = all_shares[party_rank][0] + sum(
+                all_shares[(party_rank + 1) % (nparties)]
+            )
+
+            if not value:
+                value = share_sum
+
+            elif share_sum != value:
+                raise ValueError(
+                    "Reconstruction values from all parties are not equal."
+                )
+
+        if get_shares:
+            return all_shares
+
+        return value
+
+    @staticmethod
+    def reconstruct(
+        share_ptrs: List["ReplicatedSharedTensor"],
+        get_shares: bool = False,
+        security_type: str = "semi-honest",
+    ) -> Union[torch.Tensor, List[torch.Tensor]]:
+        """Reconstruct value from shares.
+
+        Args:
+            share_ptrs (List[ReplicatedSharedTensor]): List of RSTensor pointers.
+            security_type (str): Type of security followed by protocol.
+            get_shares (bool): Retrieve only shares.
+
+        Returns:
+            reconstructed_value (torch.Tensor): Reconstructed value.
+
+        Raises:
+            ValueError: Invalid security type
+        """
+        if security_type == "malicious":
+            return ReplicatedSharedTensor.__reconstruct_malicious(
+                share_ptrs, get_shares
+            )
+
+        elif security_type == "semi-honest":
+            return ReplicatedSharedTensor.__reconstruct_semi_honest(
+                share_ptrs, get_shares
+            )
+
+        raise ValueError("Invalid security Type")
+
+    @staticmethod
+    def distribute_shares(shares: List[ShareTensor], session: Session) -> List:
+        """Distribute a list of shares.
+
+        Args:
+            shares (List[ShareTensor): list of shares to distribute.
+            session (Session): Session.
+
+        Returns:
+            List of ShareTensor.
+
+        """
+        parties = session.parties
+
+        nshares = len(parties) - 1
+
+        ptr_list = []
+        for i in range(len(parties)):
+            party_ptrs = []
```

Comment:
MAYBE : We could have it as `party_shares` as we are only appending tensor shares.

Reply:
Okay